Flutter_TDD_Architecture_Course_notes.md

```diff
@@ -1013,3 +1013,402 @@ class NetworkInfoImpl implements NetworkInfo {
+      (_) async => http.Response('Something went wrong', 404),
+    );
+    // act
+    final call = dataSource.getConcreteNumberTrivia;
+    // assert
+    expect(() => call(tNumber), throwsA(TypeMatcher<ServerException>()));
+  },
+);
+```
+
+We `implement` in the corresponding file using a if .
+
+```dart
+@override
+Future<NumberTriviaModel> getConcreteNumberTrivia(int number) async {
+  final response = await client.get(
+    'http://numbersapi.com/$number',
+    headers: {'Content-Type': 'application/json'},
+  );
+
+  if (response.statusCode == 200) {
+    return NumberTriviaModel.fromJson(json.decode(response.body));
+  } else {
+    throw ServerException();
+  }
+}
+```
+
+Now let's create two methods, one for `success` and one for `error`.
+
+```dart
+ void setUpMockHttpClientSuccess200() {
+    when(mockHttpClient.get(any, headers: anyNamed('headers'))).thenAnswer(
+      (_) async => http.Response(fixture('trivia.json'), 200),
+    );
+  }
+
+  void setUpMockHttpClientFailure404() {
+    when(mockHttpClient.get(any, headers: anyNamed('headers'))).thenAnswer(
+      (_) async => http.Response('Something went wrong', 404),
+    );
+  }
+}
+```
+
+We'll have to do the same tests but this time for a `random number`.
+
+```dart
+group('getRandomNumberTrivia', () {
+  final tNumberTriviaModel =
+      NumberTriviaModel.fromJson(json.decode(fixture('trivia.json')));
+
+  test(
+    'should preform a GET request on a URL with *random* endpoint with application/json header',
+    () {
+      //arrange
+      setUpMockHttpClientSuccess200();
+      // act
+      dataSource.getRandomNumberTrivia();
+      // assert
+      verify(mockHttpClient.get(
+        'http://numbersapi.com/random',
+        headers: {'Content-Type': 'application/json'},
+      ));
+    },
+  );
+
+  test(
+    'should return NumberTrivia when the response code is 200 (success)',
+    () async {
+      // arrange
+      setUpMockHttpClientSuccess200();
+      // act
+      final result = await dataSource.getRandomNumberTrivia();
+      // assert
+      expect(result, equals(tNumberTriviaModel));
+    },
+  );
+
+  test(
+    'should throw a ServerException when the response code is 404 or other',
+    () async {
+      // arrange
+      setUpMockHttpClientFailure404();
+      // act
+      final call = dataSource.getRandomNumberTrivia;
+      // assert
+      expect(() => call(), throwsA(TypeMatcher<ServerException>()));
+    },
+  );
+});
+```
+
+Now we implement the random function in the `number_trivia_remote_data_source.dart` file.
+
+
+```dart
+@override
+Future<NumberTriviaModel> getRandomNumberTrivia() async {
+  final response = await client.get(
+    'http://numbersapi.com/random',
+    headers: {'Content-Type': 'application/json'},
+  );
+
+  if (response.statusCode == 200) {
+    return NumberTriviaModel.fromJson(json.decode(response.body));
+  } else {
+    throw ServerException();
+  }
+}
+```
+
+Now the tests will `pass` on both sides.
```

Comment:
Please _show_ the tests passing.

Reply:
I've added the screenshot of all the tests passing.

Reply:
Thanks. 👍